### PR TITLE
feat: use DifferentiationInterface for sparse AD

### PR DESCRIPTION
**Pull request by avik-pal**
The only part where we don't use DI is structured Jacobians. That still uses SparseDiffTools

### TODO:

- [x] construct a case with structured jacobian that is small enough for testing. something block diagonal is the easiest to generate
- [x] merge https://github.com/SciML/NonlinearSolve.jl/pull/470
- [x] start removing sparsedifftools from the tutorials/docs
  - [x] docs
  - [x] tests
- [x] needs https://github.com/gdalle/SparseMatrixColorings.jl/pull/132

### CODE REVIEW

**Review by gdalle — COMMENTED** [review]
Congrats on the adaptation and thank you for trusting me with this!
I've never used LinearSolve before so take my comments with a grain of salt, but I find it needlessly complicated to mix the SparseDiffTools legacy syntax with the new [sparse API](https://sciml.github.io/ADTypes.jl/stable/#Sparse-AD) of ADTypes. Do you think it is a problem?

File: docs/Project.toml [hunk 1]
```diff
@@ -16,6 +17,7 @@ Random = "9a3f8284-a2c9-5f02-9a11-845980a1fd5c"
 SciMLBase = "0bca4576-84f4-4d90-8ffe-ffa030f20462"
 SciMLJacobianOperators = "19f34311-ddf3-4b8b-af20-060888a46c0e"
 SimpleNonlinearSolve = "727e6d20-b764-4bd8-a329-72de5adea6c7"
+SparseConnectivityTracer = "9f842d2f-2579-4b1d-911e-f412cf18a3f5"
```
[the remark this answers is not in the document]
Comment (avik-pal) [reply docs/Project.toml:20]: NonlinearSolve will always have the sparsity detection functionality, so end-users don't need to load anything. Symbolics was problematic so we did not load it, but the current ones are lightweight.

The reason I wanted SparseArrays removed from DI was to do with SimpleNonlinearSolve which needs to be as lightweight as possible.

File: docs/src/basics/sparsity_detection.md [hunk 2]
```diff
@@ -48,14 +49,18 @@ prob = NonlinearProblem(
     NonlinearFunction(nlfunc; sparsity = SymbolicsSparsityDetection()), x0)  # Remember to have Symbolics.jl loaded
```
[the remark this answers is not in the document]
Comment (avik-pal) [reply docs/src/basics/sparsity_detection.md:49]: no my bad

File: docs/src/basics/sparsity_detection.md [hunk 3]
```diff
@@ -34,10 +29,16 @@ If the `colorvec` is not provided, then it is computed on demand.
...
-    colorvec is the column colorvec, while Reverse Mode will assume that the colorvec is the
+    autodiff backend used. `ADTypes.mode(ad) isa ADTypes.ForwardMode` will assume that the
+    colorvec is the column colorvec, otherwise we will assume that the colorvec is the
     row colorvec.
 
+!!! warning
+    
+    Previously you could provide a `sparsity` argument to `NonlinearFunction` to specify
+    the jacobian prototype. However, to avoid confusion, this is now deprecated. Instead,
+    use the `jac_prototype` argument.
```
[the remark this answers is not in the document]
Comment (avik-pal) [reply docs/src/basics/sparsity_detection.md:40]: clarified. jac_prototype will do this internally so I am hesitant to tell the user to do this.

File: src/internal/jacobian.jl [hunk 4]
```diff
@@ -188,41 +197,137 @@ function (cache::JacobianCache{iip})(
...
 
-function sparsity_detection_alg(f::NonlinearFunction, ad::AbstractADType)
-    # TODO: Also handle case where colorvec is provided
-    f.sparsity === nothing && return NoSparsityDetection()
-    return sparsity_detection_alg(f, AutoSparse(ad; sparsity_detector = f.sparsity))
-end
-
-function sparsity_detection_alg(f::NonlinearFunction, ad::AutoSparse)
+function construct_concrete_adtype(f::NonlinearFunction, ad::AbstractADType)
+    @assert !(ad isa AutoSparse) "This shouldn't happen. Open an issue."
```
[the remark this answers is not in the document]
Comment (avik-pal) [reply src/internal/jacobian.jl:201]: There is another dispatch below for AutoSparse. Here it must be a dense backend

**Comment by avik-pal** [comment]
> but I find it needlessly complicated to mix the SparseDiffTools legacy syntax with the new [sparse API](https://sciml.github.io/ADTypes.jl/stable/#Sparse-AD) of ADTypes.

This preserves support for structured matrices. Once the support for that lands in SparseMatrixColorings we can just delete that check and everything will just work.

**Comment by avik-pal** [comment]
To clarify the rationale behind marking `AutoSparse` as deprecated:

Previously we had sparse backends (simple wrapper over dense backend) and we had sparsity specification inside the problem (`sparsity`, `jac_prototype`, `colorvec`). But now the AutoSparse backend holds the sparsity detection + coloring information. So this creates chances for ambiguity, say user says `sparsity = Symbolics..Detector()` and `AutoSparse(..., Tracer...Detector())` what should take precedence? We could choose to make the autodiff via the solver have higher precedence. But the other way is also perfectly valid. This mostly causes confusion without a clear benefit (at least to me)

Now what does the new-API look like? Users always provide a dense_ad to the solver. If the function has sparsity information in the form of a detector/colorvec/prototype (basically any hint that the jacobian is sparse) we construct a AutoSparse backend using that information. From DI's perspective it always sees the final autodiff that we constructed and not the one user passed in.

**Comment by gdalle** [comment]
Right, so essentially you reconstruct an `AutoSparse` from
- the backend provided to the solver
- the sparsity detector / coloring provided to the function

That seems reasonable for now, but it might lead to a breaking change when we swing back to full `AutoSparse` support?

**Comment by gdalle** [comment]
> This preserves support for structured matrices. Once the support for that lands in SparseMatrixColorings we can just delete that check and everything will just work.

Just to clarify, with structured matrices, there are two things we can optimize:

- the coloring, for which an optimal solution is often known
- the decompression, because storage is e.g. bandwise instead of columnwise

At the moment, if I understand correctly:

- the coloring is in ArrayInterface.jl, reused by SparseDiffTools.jl?
- the decompression is in these [extensions](https://github.com/JuliaDiff/FiniteDiff.jl/tree/master/ext) of FiniteDiff.jl, reused by SparseDiffTools.jl

I'll try to come up with a prototype putting everything in SparseMatrixColorings.jl

**Comment by gdalle** [comment]
On this branch you have optimized coloring and decompression for `Diagonal`, `Bidiagonal, `Tridiagonal` and `BandedMatrix`: https://github.com/gdalle/SparseMatrixColorings.jl/pull/132

**Comment by avik-pal** [comment]
> That seems reasonable for now, but it might lead to a breaking change when we swing back to full AutoSparse support?

What do you mean by full `AutoSparse` support?

I am generally against having 2 APIs to accomplish the exact same thing _unless each API provides additional disjoint benefits_. If you are referring to the coloring_algorithm, we can always rename `colorvec` to `coloring` and construct AutoSparse based on the type of `coloring`.

@ChrisRackauckas what do you think about this new API proposal for v4?

**Comment by gdalle** [comment]
> What do you mean by full AutoSparse support?

I meant that for users who pass an `AutoSparse` backend to NonLinearSolve, it might be surprising to see that the choices of `coloring_algorithm` / `sparsity_detector` inside the backend are discarded, in favor of the choices `colorvec` / `sparsity` given in the function itself?

**Comment by avik-pal** [comment]
> If you want to deprecate this way of handling sparsity, I would almost prefer throwing an error if someone gives you an AutoSparse.

Yes I agree. The function currently throws a depwarn, that will become an error in v4

**Comment by gdalle** [comment]
I added support for `BlockBandedMatrix` and `BandedBlockBandedMatrix` to the same PR. Now we have the exact same colorings as ArrayInterface, although decompression could use a little more work (but it's not terrible).

**Comment by gdalle** [comment]
What else would you need to drop SparseDiffTools completely? In terms of functionality it should be all there, performance may still lag a bit behind but honestly I'm not even sure.

**Comment by avik-pal** [comment]
see the other PR. tests pass locally for me, so we can go ahead and remove sparsedifftools. I will make some additional pushes to remove sparse diff tools from the tests and docs

**Comment by avik-pal** [comment]
Added a big table to clarify the selection mechanism https://github.com/SciML/NonlinearSolve.jl/blob/ap/sparse_di/docs/src/basics/sparsity_detection.md

**Comment by oscardssmith** [comment]
Is there anything we should be doing to pass down the sparsity info that we get down to LinearSolve? It seems like knowing that should be able to make the solve faster.

**Comment by avik-pal** [comment]
We are giving LinearSolve the exact matrix it needs to solve

**Comment by avik-pal** [comment]
We assume for `SparseMatrixCSC` jacobians the sparsity struture cannot arbitrarily change over iterations

**Comment by gdalle** [comment]
Yeah that goes without saying for me but if you prepare DI's Jacobian with a given sparsity pattern, you'll get incorrect outputs if the sparsity changes between points

**Comment by gdalle** [comment]
Fair warning: you're probably going to see an epic slowdown when computing sparse Jacobians with FiniteDiff. The reason is that, in DI, sparse Jacobians rely on pushforwards, and I don't have a native pushforward operator in FiniteDiff so I make do with a derivative closure.

**Comment by oscardssmith** [comment]
Can a native pushforward be added to FiniteDiff? it seems like it should be pretty straightforward...

**Comment by gdalle** [comment]
Knock yourselves out! I'll happily use it.

**Comment by gdalle** [comment]
In the meantime, maybe I'll try to use the sparse Jacobian inside FiniteDiff as a workaround

**Comment by oscardssmith** [comment]
https://github.com/JuliaDiff/FiniteDiff.jl/pull/191

**Comment by gdalle** [comment]
@avik-pal with the very newest versions of DifferentiationInterface (0.6.4) and SparseMatrixColorings (0.4.4), your structured tests from #470 should also pass. 
If you just want the thing to work but you're okay with slightly suboptimal colorings until https://github.com/gdalle/SparseMatrixColorings.jl/pull/139 is merged, then you can remove the dependency on a specific branch and use the latest from the registry.

File: docs/src/tutorials/large_systems.md [hunk 5]
```diff
@@ -160,8 +166,14 @@ declaration of Jacobian sparsity types. To see this in action, we can give an ex
 and `u` and call `jacobian_sparsity` on our function with the example arguments, and it will
```
[the remark this answers is not in the document]
Comment (avik-pal) [reply docs/src/tutorials/large_systems.md:167]: done

**Pull request merged**